hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/process/Signal.java

```diff
@@ -13,29 +14,45 @@
     private final Object payload;
 
     private final long executeAfterTimestamp;
+    private final long id;
 
     public enum SignalType {
         START, STOP, RETRANSMIT, UPDATE_SUBSCRIPTION, UPDATE_TOPIC, KILL,
         RESTART, RESTART_UNHEALTHY, STOP_RESTART, KILL_UNHEALTHY, CLEANUP, COMMIT
     }
 
-    private Signal(SignalType type, SubscriptionName target, Object payload, long executeAfterTimestamp) {
+    private static AtomicLong SIGNALS_COUNTER = new AtomicLong();
+
+    private Signal(SignalType type, SubscriptionName target, Object payload, long executeAfterTimestamp, long id) {
         this.type = type;
         this.target = target;
         this.payload = payload;
         this.executeAfterTimestamp = executeAfterTimestamp;
+        this.id = id;
+    }
+
+    public static Signal of(SignalType type, SubscriptionName target) {
+        return of(type, target, null);
     }
 
     public static Signal of(SignalType type, SubscriptionName target, Object payload) {
-        return new Signal(type, target, payload, -1);
+        return of(type, target, payload, -1);
     }
 
-    public static Signal of(SignalType type, SubscriptionName target) {
-        return new Signal(type, target, null, -1);
+    public static Signal of(SignalType type, SubscriptionName target, Object payload, long executeAfterTimestamp) {
+        return new Signal(type, target, payload, executeAfterTimestamp, SIGNALS_COUNTER.incrementAndGet());
     }
 
-    public static Signal of(SignalType type, SubscriptionName target, long executeAfterTimestamp) {
-        return new Signal(type, target, null, executeAfterTimestamp);
+    Signal childOf(SignalType type) {
```

Comment:
The relationship is inverted. This method should be called `child`/`offspring`/`createChild`. Consider this:

    Signal parent - Signal.of(STOP, ...);
    Signal child = parent.childOf(KILL); // feels that parent turns to a child of KILL and becomes a child in this relationship

versus:

    Signal parent - Signal.of(STOP, ...);
    Signal child = parent.offspring(KILL);